File: app/src/main/java/com/qinshou/rcvbaseadapterdemo/adapter/RcvBaseAdapter.java
package com.qinshou.rcvbaseadapterdemo.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.qinshou.rcvbaseadapterdemo.holder.BaseViewHolder;
import com.qinshou.rcvbaseadapterdemo.listener.OnItemClickListener;

import java.util.ArrayList;
import java.util.List;

/**
 * Description:RecyclerView 的适配器的基类
 * Created by 禽兽先生
 * Created on 2017/11/22
 */

public abstract class RcvBaseAdapter<T> extends RecyclerView.Adapter<BaseViewHolder> {

    private Context context;
    private int layoutId;
    private List<T> dataList = new ArrayList<>();
    private OnItemClickListener mOnItemClickListener;

    public RcvBaseAdapter(Context context, int layoutId) {
        this.context = context;
        this.layoutId = layoutId;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(context).inflate(layoutId, parent, false);
        return new BaseViewHolder(context, itemView);
    }

    @Override
    public void onBindViewHolder(final BaseViewHolder holder, int position) {
        if (mOnItemClickListener != null) {
            holder.getItemView().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int position = holder.getLayoutPosition();
                    mOnItemClickListener.onItemClick(holder, dataList.get(position), position);
                }
            });
        }
        bindViewHolder(holder, dataList.get(position), position);
    }

    @Override
    public int getItemCount() {
        return dataList == null ? 0 : dataList.size();
    }

    public abstract void bindViewHolder(BaseViewHolder holder, T itemData, int position);

    public Context getContext() {
        return context;
    }

    public int getLayoutId() {
        return layoutId;
    }

    public List<T> getDataList() {
        return dataList;
    }

    /**
     * Description:设置数据,所有数据将被替换
     * Date:2018/3/9
     *
     * @param dataList 需要添加的数据
     */
    public void setDataList(List<T> dataList) {
        this.dataList = dataList;
        notifyDataSetChanged();
    }

    /**
     * Description:添加数据
     * Date:2018/3/9
     *
     * @param dataList  需要添加的数据
     * @param isRefresh 是否是刷新,如果为 true 则会先清空当前数据再添加,为 false 则会追加
     */
    public void addDataList(List<T> dataList, boolean isRefresh) {
        if (isRefresh) {
            this.dataList.clear();
        }
        this.dataList.addAll(dataList);
        if (isRefresh) {
            notifyDataSetChanged();
        } else {
            notifyItemRangeInserted(this.dataList.size() - dataList.size(), this.dataList.size());
        }
    }

    /**
     * Description:设置 item 的点击监听器
     * Date:2018/3/9
     */
    public void setOnItemClickListener(OnItemClickListener<T> onItemClickListener) {
        this.mOnItemClickListener = onItemClickListener;
    }
}

File: rcvbaseadapter/src/main/java/com/qinshou/rcv/adapter/RcvBaseAdapter.kt
package com.qinshou.rcv.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.qinshou.rcv.listener.OnItemClickListener
import com.qinshou.rcv.listener.OnItemLongClickListener
import com.qinshou.rcv.viewholder.BaseViewHolder
import java.util.*

/**
 * Author: QinHao
 * Email:qinhao@jeejio.com
 * Date: 2021/4/14 11:14
 * Description:The base adapter of RecyclerView.If datum are the same type,please use
 * {@link RcvSimpleAdapter}.If datum are different type or datum are the same type but want to use
 * different layout,please use {@link RcvMultipleAdapter}.
 */
abstract class RcvBaseAdapter<T>(val context: Context, private val layoutId: Int) : RecyclerView.Adapter<BaseViewHolder>() {
    /**
     * Collection to save data
     */
    var dataList: MutableList<T>? = ArrayList()
        set(value) {
            if (value == null) {
                field = ArrayList()
            } else {
                field = value
            }
            notifyDataSetChanged()
        }

    /**
     * The Listener of item click
     */
    protected val onItemClickListener: OnItemClickListener<T>? = null

    /**
     * The Listener of item long click
     */
    protected val onItemLongClickListener: OnItemLongClickListener<T>? = null

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 17:36
     * Description:Do bind UI operate,children only needs to implement this method for complete the
     * adapter's tedious operate.
     */
    abstract fun bindViewHolder(holder: BaseViewHolder, itemData: T?, position: Int)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return BaseViewHolder(context, LayoutInflater.from(context).inflate(layoutId, parent, false))
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
//        if (dataList == null) {
//            return
//        }
//        if (dataList?.size == 0) {
//            return
//        }
        if (onItemClickListener != null) {
            holder.itemView.setOnClickListener(View.OnClickListener {
                onItemClickListener.onClick(holder, dataList?.get(position), position)
            })
        }
        if (onItemLongClickListener != null) {
            holder.itemView.setOnLongClickListener(OnLongClickListener {
                onItemLongClickListener.onLongClick(holder, dataList?.get(position), position)
                true
            })
        }
        bindViewHolder(holder, dataList?.get(position), position)
    }

    override fun getItemCount(): Int {
        return if (dataList == null) 0 else dataList!!.size
    }

    override fun getItemViewType(position: Int): Int {
        return super.getItemViewType(position)
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 17:39
     * Description:Add datum to collection.
     * @param dataList The datum you want to add.
     * @param isRefresh If true adapter will refresh all,false only refresh the added section
     */
    fun addDataList(dataList: List<T>?, isRefresh: Boolean = false) {
        if (dataList == null || dataList.isEmpty()) {
            return
        }
        if (isRefresh) {
            this.dataList?.clear()
        }
        this.dataList?.addAll(dataList)
        if (isRefresh) {
            notifyDataSetChanged()
        } else {
            notifyItemRangeInserted(this.dataList!!.size - dataList.size, this.dataList!!.size)
        }
    }
}